.orchestra/ci/ci-hook/app.py

```diff
@@ -89,77 +159,309 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         assert not self.user.impersonationtokens.get(impersonation_token_id).active
 
 
-def trigger_ci(username, repo_url, ref, before, after):
+def impersonate(admin_gl, username):
+    matching = admin_gl.users.list(username=username)
+    if len(matching) != 1:
+        raise Exception("Unexpected number of users matching " + username)
+    user_id = matching[0].id
+
+    user = admin_gl.users.get(user_id)
+    return ImpersonationToken(user)
+
+
+def trigger_ci(username, repo_url, base_repo_url, ref, status_update_metadata: Optional[dict] = None):
     # Ignore pushes by the CI itself
     if username == ci_user:
         return
 
     admin_gl = gitlab.Gitlab(GITLAB_URL, private_token=ADMIN_TOKEN)
     admin_gl.auth()
 
-    matching = admin_gl.users.list(username=username)
-    if len(matching) != 1:
-        raise Exception("Unexpected number of users matching " + username)
-    user_id = matching[0].id
+    try:
+        with impersonate(admin_gl, username) as token:
+            user_gl = gitlab.Gitlab(GITLAB_URL, private_token=token)
+            user_gl.auth()
+            project = user_gl.projects.get(PROJECT_ID)
+
+            is_anonymous = username == default_user
+
+            variables = {
+                "TARGET_COMPONENTS_URL": repo_url,
+                "TARGET_COMPONENTS": " ".join(is_anonymous and default_user_target_components or target_components),
+                "PUSHED_REF": ref,
+                "ORCHESTRA_CONFIG_REPO_HTTP_URL": ORCHESTRA_CONFIG_REPO_HTTP_URL,
+                "ORCHESTRA_CONFIG_REPO_SSH_URL": ORCHESTRA_CONFIG_REPO_SSH_URL,
+            }
+
+            if status_update_metadata is not None:
+                variables["REVNG_CI_STATUS_UPDATE_METADATA"] = json.dumps(status_update_metadata)
+
+            if username in allowed_to_push and base_repo_url == repo_url:
+                variables["BASE_USER_OPTIONS_YML"] = pusher_user_options
+                variables["SSH_PRIVATE_KEY"] = revng_push_ci_private_key
+                variables["PUSH_CHANGES"] = "1"
+            else:
+                tpl_params = {
+                    "clone_namespace": "/".join(repo_url.split("/")[:-1]),
+                    "private_sources": "",
+                    "private_bin_archives": ""
+                }
+                if not is_anonymous:
+                    # Placeholders are replaced in the shell script since we don't want to intentionally introduce shell
+                    # injection vulnerabilities.
+                    tpl_params.update({
+                        "private_sources": "- private: %PRIVATE_SOURCES_PLACEHOLDER%",
+                        "private_bin_archives": "- private: %PRIVATE_BIN_ARCHIVES_PLACEHOLDER%"
+                    })
+
+                variables["BASE_USER_OPTIONS_YML"] = \
+                    string.Template(pull_request_user_options).substitute(tpl_params)
```

Comment:
Why do this only if `username not in allowed_to_push`?

Reply:
The full condition is `username in allowed_to_push and base_repo_url == repo_url` and it's important.

If it is true it means that you're triggering the CI from **a push to the repo itself** (not a fork) and that you're trying to promote a branch. So the SSH user_options should be used.

If it is false it means that it is either being launched from a fork or from another user. In this case we want to add the user namespace to the options, and we want internal users to be able to access the private repos but by using the internal impersonation token authentication, not the SSH keys with push access.

Reply:
Gotcha, docs docs docs.